velox/exec/ExchangeClient.cpp

```diff
@@ -170,74 +174,42 @@ void ExchangeClient::request(const RequestSpec& requestSpec) {
   }
 }
 
-int32_t ExchangeClient::countPendingSourcesLocked() {
-  int32_t numPending = 0;
-  for (auto& source : sources_) {
-    if (source->isRequestPendingLocked()) {
-      ++numPending;
-    }
-  }
-  return numPending;
-}
-
-int64_t ExchangeClient::getAveragePageSize() {
-  auto averagePageSize =
-      std::min<int64_t>(maxQueuedBytes_, queue_->averageReceivedPageBytes());
-  if (averagePageSize == 0) {
-    averagePageSize = 1 << 20; // 1 MB.
+std::vector<ExchangeClient::RequestSpec>
+ExchangeClient::pickSourcesToRequestLocked() {
+  if (closed_) {
+    return {};
   }
-
-  return averagePageSize;
-}
-
-int32_t ExchangeClient::getNumSourcesToRequestLocked(int64_t averagePageSize) {
-  // Figure out how many more 'averagePageSize' fit into 'maxQueuedBytes_'.
-  // Make sure to leave room for 'numPending' pages.
-  const auto numPending = countPendingSourcesLocked();
-
-  auto numToRequest = std::max<int32_t>(
-      1, (maxQueuedBytes_ - queue_->totalBytes()) / averagePageSize);
-  if (numToRequest <= numPending) {
-    return 0;
+  std::vector<RequestSpec> requestSpecs;
+  while (!emptySources_.empty()) {
+    auto& source = emptySources_.front();
+    bool ok = source->shouldRequestLocked();
+    VELOX_CHECK(ok);
+    requestSpecs.push_back({std::move(source), 0});
+    emptySources_.pop();
   }
-
-  return numToRequest - numPending;
-}
-
-void ExchangeClient::pickSourcesToRequestLocked(
-    RequestSpec& requestSpec,
-    int32_t numToRequest,
-    std::queue<std::shared_ptr<ExchangeSource>>& sources) {
-  while (requestSpec.sources.size() < numToRequest && !sources.empty()) {
-    auto& source = sources.front();
-    if (source->shouldRequestLocked()) {
-      requestSpec.sources.push_back(source);
+  int64_t availableSpace =
+      maxQueuedBytes_ - queue_->totalBytes() - totalPendingBytes_;
```

Comment:
> uint64_t totalBytes() const {

I remember seeing a bug here because queue_->totalBytes() is an unsigned int and that caused the whole computation to be done in unsigned ints leading to some wrong results. Is this not happening ?

Reply:
Because the final result is casted back to signed integer so we can check if it is positive correctly.  The addition and subtraction for signed and unsigned integers are same.  Only the comparison could be different.

Reply:
I see. This code works because you explicitly specify the type for availableSpace as int64_t and not using auto. If this code was changed to use auto, it would stop working. Would it be possible to add an explicit cast `(int64_t) queue_->totalBytes()` and use auto to prevent accidental changes to auto from breaking things. Or, perhaps, check that changing to 'auto' will result in a test failure. Thanks.

```
  const uint64_t x = 123;
  const int64_t y = 100;

  const int64_t a = y - x;
  const auto b = y - x;

  ASSERT_EQ(a, b); // -23 vs. 18446744073709551593
```

Reply:
I think we should just fix the signature of `totalBytes()` to return `int64_t`.  We prefer signed integer whenever it need arithmetic calculation for this exact reason.

Reply:
Makes sense. Thanks.